deploy/examples/create-external-cluster-resources.py

```diff
@@ -93,12 +93,12 @@ def _init_cmd_output_map(self):
-        self.cmd_output_map['''{"caps": ["mon", "profile rbd", "osd", "profile rbd"], "entity": "client.csi-rbd-node", "format": "json", "prefix": "auth get-or-create"}'''] = '''[{"entity":"client.csi-rbd-node","key":"AQBOgrNeHbK1AxAAubYBeV8S1U/GPzq5SVeq6g==","caps":{"mon":"profile rbd","osd":"profile rbd"}}]'''
-        self.cmd_output_map['''{"caps": ["mon", "profile rbd", "mgr", "allow rw", "osd", "profile rbd"], "entity": "client.csi-rbd-provisioner", "format": "json", "prefix": "auth get-or-create"}'''] = '''[{"entity":"client.csi-rbd-provisioner","key":"AQBNgrNe1geyKxAA8ekViRdE+hss5OweYBkwNg==","caps":{"mgr":"allow rw","mon":"profile rbd","osd":"profile rbd"}}]'''
-        self.cmd_output_map['''{"caps": ["mon", "allow r", "mgr", "allow rw", "osd", "allow rw tag cephfs *=*", "mds", "allow rw"], "entity": "client.csi-cephfs-node", "format": "json", "prefix": "auth get-or-create"}'''] = '''[{"entity":"client.csi-cephfs-node","key":"AQBOgrNeENunKxAAPCmgE7R6G8DcXnaJ1F32qg==","caps":{"mds":"allow rw","mgr":"allow rw","mon":"allow r","osd":"allow rw tag cephfs *=*"}}]'''
-        self.cmd_output_map['''{"caps": ["mon", "allow r", "mgr", "allow rw", "osd", "allow rw tag cephfs metadata=*"], "entity": "client.csi-cephfs-provisioner", "format": "json", "prefix": "auth get-or-create"}'''] = '''[{"entity":"client.csi-cephfs-provisioner","key":"AQBOgrNeAFgcGBAAvGqKOAD0D3xxmVY0R912dg==","caps":{"mgr":"allow rw","mon":"allow r","osd":"allow rw tag cephfs metadata=*"}}]'''
-        self.cmd_output_map['''{"caps": ["mon", "allow r", "mgr", "allow rw", "osd", "allow rw tag cephfs metadata=*"], "entity": "client.csi-cephfs-provisioner-openshift-storage", "format": "json", "prefix": "auth get-or-create"}'''] = '''[{"entity":"client.csi-cephfs-provisioner-openshift-storage","key":"BQBOgrNeAFgcGBAAvGqKOAD0D3xxmVY0R912dg==","caps":{"mgr":"allow rw","mon":"allow r","osd":"allow rw tag cephfs metadata=*"}}]'''
-        self.cmd_output_map['''{"caps": ["mon", "allow r", "mgr", "allow rw", "osd", "allow rw tag cephfs metadata=myfs"], "entity": "client.csi-cephfs-provisioner-openshift-storage-myfs", "format": "json", "prefix": "auth get-or-create"}'''] = '''[{"entity":"client.csi-cephfs-provisioner-openshift-storage-myfs","key":"CQBOgrNeAFgcGBAAvGqKOAD0D3xxmVY0R912dg==","caps":{"mgr":"allow rw","mon":"allow r","osd":"allow rw tag cephfs metadata=myfs"}}]'''
+        self.cmd_output_map['''{"caps": ["mon", "profile rbd, profile simple-rados-client-with-blocklist", "osd", "profile rbd"], "entity": "client.csi-rbd-node", "format": "json", "prefix": "auth get-or-create"}'''] = '''[{"entity":"client.csi-rbd-node","key":"AQBOgrNeHbK1AxAAubYBeV8S1U/GPzq5SVeq6g==","caps":{"mon":"profile rbd, profile simple-rados-client-with-blocklist","osd":"profile rbd"}}]'''
```

Comment:
will this work for the octopus release? @parth-gr have you validated that with this user now you are able to execute blocklist commands?

Reply:
with below caps i was not able to do blocklisting

```
client.csi-rbd-provisioner
	key: AQCCkeVhb8xhHhAAfh7GZ3ilYmSW8dqh7GzEIw==
	caps: [mgr] allow rw
	caps: [mon] profile rbd
	caps: [osd] profile rbd

sh-4.4$ ceph osd blocklist add 172.17.0.1 --user csi-rbd-provisioner --key=AQCCkeVhb8xhHhAAfh7GZ3ilYmSW8dqh7GzEIw==
Error EACCES: access denied
```
but with the below user am able to do it.
```
sh-4.4$ ceph version
ceph version 16.2.6 (ee28fb57e47e9f88813e24bbf4c14496ca299d31) pacific (stable)

ceph auth get-or-create client.kubernetes mon 'profile rbd' mon 'allow rw' osd 'profile rbd' mgr 'allow rw'
[client.kubernetes]
	key = AQB3kuVhvCxvKBAAjvdlZ3b41lMs2z/uTsEpGA==
sh-4.4$ ceph osd blocklist add 172.17.0.1 --user kubernetes --key=AQB3kuVhvCxvKBAAjvdlZ3b41lMs2z/uTsEpGA==         
blocklisting 172.17.0.1:0/0 until 2022-01-17T17:00:09.782294+0000 (3600 sec)
```

I don't know these are the exact caps required. @idryomov can you please confirm the caps required to execute blocklist commands?

Reply:
We run `ceph osd blocklist add <NODE_IP>` for Octopus(v15.2.x) or older
And 
ceph osd blocklist add <NODE_IP> for pacific

Reply:
> with below caps i was not able to do blocklisting
> 
> ```
> client.csi-rbd-provisioner
> 	key: AQCCkeVhb8xhHhAAfh7GZ3ilYmSW8dqh7GzEIw==
> 	caps: [mgr] allow rw
> 	caps: [mon] profile rbd
> 	caps: [osd] profile rbd
> 
> sh-4.4$ ceph osd blocklist add 172.17.0.1 --user csi-rbd-provisioner --key=AQCCkeVhb8xhHhAAfh7GZ3ilYmSW8dqh7GzEIw==
> Error EACCES: access denied
> ```

I don't see the caps are being updated here with your client auth,

Its should be like 
```
client.csi-rbd-provisioner
        key: AQDoL+Vhr3ZcMhAA+HgaAFVQoG7tV/1HwxqfTQ==
        caps: [mgr] allow rw
        caps: [mon] profile rbd, profile simple-rados-client-with-blocklist
```

to work

Reply:
@Madhu-1 you can see here the caps we use for osd blocklisting https://github.com/ceph/ceph/blob/master/doc/rados/api/libcephsqlite.rst#user

Reply:
>
ceph auth get-or-create client.kubernetes mon 'profile rbd' mon 'allow rw' osd 'profile rbd' mgr 'allow rw'
[client.kubernetes]
	key = AQB3kuVhvCxvKBAAjvdlZ3b41lMs2z/uTsEpGA==
sh-4.4$ ceph osd blocklist add 172.17.0.1 --user kubernetes --key=AQB3kuVhvCxvKBAAjvdlZ3b41lMs2z/uTsEpGA==         
blocklisting 172.17.0.1:0/0 until 2022-01-17T17:00:09.782294+0000 (3600 sec)

see with the above caps am able to execute blocklisting i don't have the profile you mentioned here.

Reply:
Thanks @idryomov ,  but with mon caps i getting error

```
client.csi-rbd-provisioner
	key: AQCCkeVhb8xhHhAAfh7GZ3ilYmSW8dqh7GzEIw==
	caps: [mgr] allow rw
	caps: [mon] profile rbd
	caps: [osd] profile rbd

sh-4.4$ ceph osd blocklist add 172.17.0.1 --user csi-rbd-provisioner --key=AQCCkeVhb8xhHhAAfh7GZ3ilYmSW8dqh7GzEIw==
Error EACCES: access denied
sh-4.4$ ceph version
ceph version 16.2.6 (ee28fb57e47e9f88813e24bbf4c14496ca299d31) pacific (stable)
```

Reply:
You need to specify the nonce as well: `ceph osd blocklist add 172.17.0.1/123` should work.  This restriction comes from `profile rbd` and `profile simple-rados-client-with-blocklist` behaves the same.

Reply:
@idryomov Thankx you for the update, I see my current changes work perfectly as tested, 
But as you mentioned in some places it's redundant so checked it works fine too when removing the overlapping code,

Can you share the doc that explains this? For now, I was using this particular write up https://github.com/ceph/ceph/blob/master/doc/rados/api/libcephsqlite.rst#user

Reply:
@idryomov we need IP-based (future CIDR range) blocklisting as we need to block all clients from the dead cluster for the Metro DR use case. with the below user am able to blocklist the entire IP address. 

```bash
sh-4.4$ ceph auth get-or-create client.kubernetes mon 'profile rbd' mon 'allow rw' osd 'profile rbd' mgr 'allow rw'
[client.kubernetes]
	key = AQB3kuVhvCxvKBAAjvdlZ3b41lMs2z/uTsEpGA==
sh-4.4$ ceph osd blocklist add 172.17.0.1 --user kubernetes --key=AQB3kuVhvCxvKBAAjvdlZ3b41lMs2z/uTsEpGA==         
blocklisting 172.17.0.1:0/0 until 2022-01-17T17:00:09.782294+0000 (3600 sec)
```

Reply:
> Can you share the doc that explains this? For now, I was using this particular write up https://github.com/ceph/ceph/blob/master/doc/rados/api/libcephsqlite.rst#user

I'm not sure if it's clearly explained anywhere.  In a nutshell, profiles are bundles of individual capabilities, hand-picked for certain use.  For example, `profile rbd` includes the blocklisting capability because RBD exclusive-lock machinery needs to be able to blocklist dead peers.

Both `profile rbd` and `profile simple-rados-client-with-blocklist` include `allow r` capability.  The example you linked confirms this: it's just `ceph auth get-or-create client.X mon 'profile simple-rados-client-with-blocklist' ...`, there is no `allow r` there.

Reply:
```bash
ceph osd blocklist add 172.17.0.1:0/0 --user csi-rbd-provisioner --key=AQCCkeVhb8xhHhAAfh7GZ3ilYmSW8dqh7GzEIw==
```
Now am able to blocklist it.

@idryomov Thanks for taking the time to explain this one. as we already have `mon 'profile rbd'` caps for the cephcsi rbd provisioner user we don't need the changes done in this PR to run the blocklisting commands?

Reply:
For adding blocklist entries, you don't -- `mon 'profile rbd'` covers it.  If you also want to remove blocklist entries (usually a bad idea), you would need _other_ changes (i.e. not what is in this PR).

Reply:
@idryomov you made a good point that is missing here, Yes we do have the requirement of removing the blocklisting as part of FailBack operation once the cluster is recovered. can you please let us know what are the caps required for that?

cc @raghavendra-talur

Reply:
> @idryomov Thanks for taking the time to explain this one. as we already have `mon 'profile rbd'` caps for the cephcsi rbd provisioner user we don't need the changes done in this PR to run the blocklisting commands?

@Madhu-1 so do we need changes for `cephFSProvisioner`, I see they don't include 'profile rbd'?

Reply:
> @idryomov you made a good point that is missing here, Yes we do have the requirement of removing the blocklisting as part of FailBack operation once the cluster is recovered. can you please let us know what are the caps required for that?
> 
> cc @raghavendra-talur

According to me `allow rw` provide it, but a not a good idea to use it from a security point of view.

Reply:
> Yes we do have the requirement of removing the blocklisting as part of FailBack operation once the cluster is recovered. can you please let us know what are the caps required for that?

In that case, append `mon 'allow command "osd blocklist"'`.  I'd like to reiterate that removing blocklist entries should be done with great caution.

Reply:
The [ceph-csi implementation](https://github.com/ceph/ceph-csi/pull/2738) of the same involves reading a CIDR block and converting it into IP range and blocklist them one by one.
In the current approach, as @Madhu-1 suggested, we directly use IP to blocklist and are missing the `ip:port/nonce` format.
@idryomov  As per your suggestion :
`The full syntax is ip:port/nonce. Specifying port = 0 and nonce = 0 should blocklist the entire IP address.`

We can update the logic to add `:0/0` to single IP addresses while blocklisting to be in sync with the rook capabilities. Will that be fine? @idryomov @Madhu-1

Reply:
Specifying the port and the nonce is more explicit (and correct IMO), but since you want to be able to remove blocklist entries and would be using `mon 'allow command "osd blocklist"'` for that, you can continue specifying just the IP.  The syntax restriction comes from profiles -- with `mon 'allow command "osd blocklist"'` it isn't enforced.

Reply:
Thanks for confirming @idryomov :)